Momentum budget terms multiplied by factional layer thicknesses

hmkhatri opened the pull request:
I have made changes to save momentum budget terms (PFu/PFv, CAu/CAv) multiplied by fractional layer thicknesses as diagnostics. These would be useful in computing depth-integrated budgets. 
Fractional thicknesses are returned in the btstep subroutine call in the MOM_dynamics_split_RK2 file. These are then used to compute required diagnostics.

gustavo-marques committed:
Adds the deterministic part of the Stanley param.

adcroft committed:
Re-wrote Stanley parameterization using vertical weights

- This now calculates the gradient of SGS temperature variance
  using the same discretization as used for the gradient of
  density along coordinate surfaces.
- Added run-time coefficient for Stanley parameterization
- Fixed openmp directives.
- Alters halo over which vert_fill_TS() is called.
- Add Stanley parameter to tc2 to test new code.

nikizadehgfdl committed:
This is the minimum required update to use 2020.01 FMS & FMScoupler

- Adding domain position arguments to diag_axis_init calls

adcroft committed:
Adds the Stanley version of EOS

- Stanley et al., 2020, adds the Brankart modification to volume mean
  density via linear corrections involving SGS sample variances and
  covariances of T and S. This commit adds the new interfaces that
  allow a call to calculate_density to include the variances and
  covariance as arguments. This correction sits above the particular
  EOS and thus can use any EOS so long as second derivatives are
  available within the EOS module.

adcroft committed:
Renamed pressure_gradient_plm() to TS_PLM_edge_values()

- The routines pressure_gradient_plm() and pressure_gradient_ppm()
  were poorly named and had comments referring to the pressure gradient
  calculation even though the only calculate edge values in the vertical
  for T/S using ALE functions. The routines are actually more general
  and can be used outside of the PGF. The comments have been shortened
  and no longer refer to the PGF.

adcroft committed:
Break out density integrals into a new module

- The integrals of density routines used (mostly) by the PGF calculation
  were part of MOM_EOS. Originally, when writing the analytic FV PGF paper,
  this was the right place to put the integrals. The additional variants
  using the ALE reconstruction functions mean that it is cleaner to have
  these routines sit in a layer above EOS and ALE.

adcroft committed:
Renamed ppoly_E to edge_values for a bit of clarity

- ppoly_E meant something to someone a while ago but we felt it
  would be better to clean up the ALE APIs. This is a pre-cursor
  to switching to a more precise description of reconstructions.

adcroft committed:
Cleaned up unused use statements

adcroft committed:
Tidied up logic for EOS_QUADRATURE

- Undid nesting of if / select statemnent
- Changed FATAL messages

adcroft committed:
Removed find_depth_of_pressure_in_cell() from MOM_EOS

- Function had been moved to density_integrals already but not
  deleted
- Also ordered public statements in MOM_EOS to help find things.

adcroft committed:
Reordered function in MOM_density_integrals

- Order was frustratingly illogical

adcroft committed:
Use SZ macros in declarations

- When converting from the two hor_index types to one (HII and HIO
  became HI) I retained the HI%isd:... declaration statements. We
  normally use `SZI_(G)` or `SZI_(HI)` so this switches to that style.

adcroft committed:
Implemented elemental PLM functions

- This adds elemental functions for cellwise operations within the PLM
  construction procedure which allows the operations to be accessed
  from outside of the ALE functions on different array shapes but recover
  bitwise identical results.
- The older subroutines now use the functions and some optimizations were
  obtained in the process.

adcroft committed:
Add time for unit_tests

- Out of curiosity I want to be sure these weren't implicated in the long
  initialization cost.

adcroft committed:
Fixed index capitalization in int_density_dz_generic_plm()

- Soft index convention was not properly implemented in original PLM
  density integrals. Fixed to avoid confusion.

adcroft committed:
Re-used pre-computed weights in int_density_dz_generic_plm()

- Probably makes no difference but we re-computed weights repeatedly
  even though we'd pre-computed them (and used only once).
  - Reduces number of local variables.

adcroft committed:
Adds the PPM form of PGF by quadrature

- Removes (now) unused functions for quadrature and polynomial evaluation
- Has been tested by setting the logical use_PPM=.false.
  - reproduces PLM mode bitwise

hmkhatri committed:
New diagnostics for barotropic momentum budget calculations are created. In these, differet budget terms multiplied by fractional layer thicknesses are saved. Thus, these terms can be added over the whole to obtain the barotropic momentum budget. 'btstep' subroutine in MOM_barotropic module is modified to return fractional thicknesses. Pressure force accelaration multiplied by fractional thickness as diagnostics are added in this commit.

hmkhatri committed:
More thickness weighted diagnostics

ashao commented on src/core/MOM_dynamics_split_RK2.F90:
```diff
@@ -306,6 +308,15 @@ subroutine step_MOM_dyn_split_RK2(u, v, h, tv, visc, Time_local, dt, forces, p_s
+  real, dimension(SZIB_(G),SZJ_(G),SZK_(G)) :: hf_CFu ! Zonal Coriolis force accel. x fract. thickness [L T-2 ~> m s-2].
+  real, dimension(SZI_(G),SZJB_(G),SZK_(G)) :: hf_CFv ! Merdional Coriolis force accel. x fract. thickness [L T-2 ~> m s-2].
+
+  real, dimension(SZIB_(G),SZJ_(G),SZK_(G)) :: hfrac_u
```
I'd suggest naming these variables something like `diag_hfrac_u` and `diag_hfrac_v` since they are not used prognostically.

ashao commented on src/core/MOM_dynamics_split_RK2.F90:
```diff
@@ -860,6 +881,20 @@ subroutine step_MOM_dyn_split_RK2(u, v, h, tv, visc, Time_local, dt, forces, p_s
   if (CS%id_u_BT_accel > 0) call post_data(CS%id_u_BT_accel, CS%u_accel_bt, CS%diag)
   if (CS%id_v_BT_accel > 0) call post_data(CS%id_v_BT_accel, CS%v_accel_bt, CS%diag)
 
+  ! Diagnostics for terms mutiplied by fractional thicknesses
```
Typo: `multiplied`

ashao commented on src/core/MOM_dynamics_split_RK2.F90:
```diff
@@ -860,6 +881,20 @@ subroutine step_MOM_dyn_split_RK2(u, v, h, tv, visc, Time_local, dt, forces, p_s
 
+  ! Diagnostics for terms mutiplied by fractional thicknesses
+  if (CS%id_hfPFu > 0) then
+    do k=1,nz ; do j=js,je ; do I=Isq,Ieq
```
These diagnostics aren't at q-points. Use the `is` and `js` instead

hmkhatri replied:
CS%PFu is at u points and CS%PFv is at v points. I am following the approach of the piece of code in lines 455-462
```
  do k=1,nz
    do j=js,je ; do I=Isq,Ieq
      u_bc_accel(I,j,k) = (CS%CAu(I,j,k) + CS%PFu(I,j,k)) + CS%diffu(I,j,k)
    enddo ; enddo
    do J=Jsq,Jeq ; do i=is,ie
      v_bc_accel(i,J,k) = (CS%CAv(i,J,k) + CS%PFv(i,J,k)) + CS%diffv(i,J,k)
    enddo ; enddo
  enddo
```

ashao replied:
Hrmm that's a little weird to me as well as to why the accelerations are using the q-point indices. I think they probably end up being the the 'correct ones' but maybe that's something to clarify on the Slack channel.

hmkhatri replied:
I am assuming that indices for acceleration terms should match du/dt and dv/dt points, which are q points.

ashao replied:
q-points should be where the divergence or curl is calculated which is why it's weird to me why the accelerations are using the q-point indices:
https://mom6.readthedocs.io/en/dev-gfdl/api/generated/pages/Horizontal_indexing.html

hmkhatri replied:
Sorry, I mixed u/v-points with q points. We want acceleration terms to be on u/v points.

`do j=js,je ; do I=Isq,Ieq`
Isn't this loop at u-points?

`do J=Jsq,Jeq ; do i=is,ie`
and this at v-points?

I think these are correct.

For q-points, the loops would look like
`do J=Jsq,Jeq ; do I=Isq,Ieq`

ashao replied:
I went and looked at the definitions of each of the indices (instead of just relying on faulty memory). Long story short, those loop indices are correct.

The `Isq` and `Jsq` variables are assigned based off the `IscB` and `JscB` indices which if you were looping over both would mean that you're at the top-right vertex of the cell, using only one of them as a loop index means that you're at the velocity point.

ashao commented on src/core/MOM_dynamics_split_RK2.F90:
```diff
@@ -746,6 +757,16 @@ subroutine step_MOM_dyn_split_RK2(u, v, h, tv, visc, Time_local, dt, forces, p_s
     call check_redundant("u_accel_bt ", CS%u_accel_bt, CS%v_accel_bt, G)
   endif
 
+  ! save fractional thickness in local variables
```
Remove commented lines

ashao commented on src/core/MOM_dynamics_split_RK2.F90:
```diff
@@ -1228,6 +1263,10 @@ subroutine initialize_dyn_split_RK2(u, v, h, uh, vh, eta, Time, G, GV, US, param
       'Zonal Pressure Force Acceleration', 'm s-2', conversion=US%L_T2_to_m_s2)
   CS%id_PFv = register_diag_field('ocean_model', 'PFv', diag%axesCvL, Time, &
       'Meridional Pressure Force Acceleration', 'm s-2', conversion=US%L_T2_to_m_s2)
+  CS%id_hfPFu = register_diag_field('ocean_model', 'hf_PFu', diag%axesCuL, Time, &
```
I think the new diagnostics probably should have `v_extensive = .true.` because they are thickness weighted.

hmkhatri replied:
I have made this change but need to confirm if` v_extensive = .true.` is correct in this case.

ashao reviewed (changes requested):
Looks good overall. Glad it seemed a bit easier than originally feared. Couple of suggestions to address. The main one that I'm not sure about is whether these diagnostics are vertically extensive. I think they are, but can you confirm? If so, then `v_extensive = .true.` should be set when the diagnostics are registered.

hmkhatri committed:
Implemented Andrew Shao's suggestions

hmkhatri committed:
More fractional thickness multiplied diagnostics

hmkhatri committed:
Some barotropic diagnostics obtained from fractional thickness multiplied momentum budget terms

hmkhatri committed:
Removed trailing spaces

adcroft committed:
Fixed line length issue in documentation

adcroft committed:
Fixed openmp in MOM_ALE.F90

- Use of new PLM functions led to out of date directives

adcroft committed:
Updated to int_density_dz_generic_ppm() to use k and tv arguments

- Now passing in 3D structures and tv so that we can access other fields
  in tv%.
- Corrected calculation of curvatures s6, t6 for interpolated interior
  line integrals.
- Tested with use_PPM=.false. and PLM edge values.
  - Using PLM edge values with use_PPM=.true. does give different answers
    because t6,s6 are non-zero albeit tiny. This is due to FP truncation
    errors.

kshedstrom committed:
Merge remote-tracking branch 'gfdl/dev/gfdl' into dev/esmg

adcroft committed:
Renamed MOM_PressureGradient_AFV

- Since the majority of code is not about the analytic integration
  I thought it time to rename this module

adcroft committed:
After renaming module, change _AFV_ to _FV_

- Continuation of rename of module

adcroft committed:
Removed conditional scaling from int_density_dz_generic_ppm()

- This routine was suffering from if's inside loops. I've removed this
  one since it was just obfuscation, pretending to care about performance
  and actually reducing it.

adcroft committed:
Adds SGS variance to tv and adds Brankart effect to PGF

- Makes use of the Stanley equation of state to include effects of
  SGS temperature variance, salinity variance and T-S covariance.

adcroft committed:
Added STANLEY_PRM_DET_COEFF run time parameter

- Renamed STANLEY_DET_COEFF to STANLEY_PRM_DET_COEFF to indicate
  this is for the Stanley parameterization as opposed to a related
  approach by Stanley for implementing the Brankart PGF correction.

adcroft committed:
Adds the T variance Stanley component to PGF

- PGF_STANLEY_T2_DET_COEFF>=0. now adds the temperature variance
  contribution to the Brankart correction using the Stanley linearization
  of the EOS and parameterization of SGS variance.

hmkhatri committed:
All fractional-thickness multiplied diagnostics implemented

adcroft committed:
Replaced copies of GSW files with links

- We always intended to use links so that we were not maintaining code
  that did not belong to us. `listpaths` used to not work as expected
  so Niki had placed copies here when adding TEOS10. Using links also
  avoids any license conflicts.

adcroft committed:
Moved pkg/MOM6_DA_hook to config_src/external

- Removed two git submodules and added bare interfaces to
  config_src/exernal/ODA_hooks/
- Added some missing documentation

adcroft committed:
Corrected documentation errors

- Updated Doxygen configuration to find files in config_src/external

adcroft committed:
Use FMS tag 2019.01.02 in .testing

- This version of FMS is a prerequisite to updating calls to FMS ready to
  try FMS2 I/O behind NOAA-GFDL/SIS2#117 and #1033

kshedstrom committed:
Merge remote-tracking branch 'gfdl/dev/gfdl' into dev/esmg

adcroft committed:
Nullified new pointers in tv

- After starting with allocatables and switching to pointers
  I'd forgotten to nullify them.

MJHarrison-GFDL committed:
Doxygen for DA hooks

hmkhatri committed:
define diagnostic variables as pointers

hmkhatri committed:
Merge branch 'dev/gfdl' into diag_TW

ashao commented on src/core/MOM_dynamics_split_RK2.F90:
```diff
@@ -777,7 +805,7 @@ subroutine step_MOM_dyn_split_RK2(u, v, h, tv, visc, Time_local, dt, forces, p_s
   call vertvisc_coef(u, v, h, forces, visc, dt, G, GV, US, CS%vertvisc_CSp, CS%OBC)
   call vertvisc(u, v, h, forces, visc, dt, CS%OBC, CS%ADp, CS%CDp, G, GV, US, &
-                CS%vertvisc_CSp, CS%taux_bot, CS%tauy_bot,waves=waves)
+                CS%vertvisc_CSp, CS%taux_bot, CS%tauy_bot,waves=waves, hfrac_u=CS%diag_hfrac_u, hfrac_v=CS%diag_hfrac_v)
```
Check to make sure that this is under 120 characters

hmkhatri replied:
Sure, I will check this. Otherwise, I will add one more line.

ashao commented on src/diagnostics/MOM_diagnostics.F90:
```diff
@@ -257,6 +264,14 @@ subroutine calculate_diagnostic_fields(u, v, h, uh, vh, tv, ADp, CDp, p_surf, &
 
   call calculate_derivs(dt, G, CS)
 
+  ! Diagnostics for terms multiplied by fractional thicknesses
+  do j=js,je ; do I=Isq,Ieq
+    hf_du_dt_2d(I,j) = 0.0
+  enddo ; enddo
+  do J=Jsq,Jeq ; do i=is,ie
+    hf_dv_dt_2d(i,J) = 0.0
+  enddo ; enddo
+
```
Most of the time explicit loops are preferred, but for initializing arrays like this, the style guide says you could have done `hf_du_dt_2d(:,:)=0.`

hmkhatri replied:
Thanks. This is much shorter. I will use this form in all arrays.

ashao commented on src/parameterizations/lateral/MOM_hor_visc.F90:
```diff
@@ -2016,6 +2064,19 @@ subroutine hor_visc_init(Time, G, US, param_file, diag, CS, MEKE)
   CS%id_diffv = register_diag_field('ocean_model', 'diffv', diag%axesCvL, Time, &
       'Meridional Acceleration from Horizontal Viscosity', 'm s-2', conversion=US%L_T2_to_m_s2)
 
+  CS%id_hf_diffu = register_diag_field('ocean_model', 'hf_diffu', diag%axesCuL, Time, &
+      'Thickness-weighted Zonal Acceleration from Horizontal Viscosity', 'm s-2', v_extensive=.true., conversion=US%L_T2_to_m_s2)
+  if (CS%id_hf_diffu > 0) call safe_alloc_ptr(CS%hf_diffu,G%IsdB,G%IedB,G%jsd,G%jed,G%ke)
+  CS%id_hf_diffv = register_diag_field('ocean_model', 'hf_diffv', diag%axesCvL, Time, &
+      'Thickness-weighted Meridional Acceleration from Horizontal Viscosity', 'm s-2', v_extensive=.true., conversion=US%L_T2_to_m_s2)
+  if (CS%id_hf_diffv > 0) call safe_alloc_ptr(CS%hf_diffv,G%isd,G%ied,G%JsdB,G%JedB,G%ke)
+  CS%id_hf_diffu_2d = register_diag_field('ocean_model', 'hf_diffu_2d', diag%axesCu1, Time, &
+      'Barotropic Thickness-weighted Zonal Acceleration from Horizontal Viscosity', 'm s-2', conversion=US%L_T2_to_m_s2)
+  if (CS%id_hf_diffu_2d > 0) call safe_alloc_ptr(CS%hf_diffu_2d,G%IsdB,G%IedB,G%jsd,G%jed)
+  CS%id_hf_diffv_2d = register_diag_field('ocean_model', 'hf_diffv_2d', diag%axesCv1, Time, &
+      'Barotropic Thickness-weighted Meridional Acceleration from Horizontal Viscosity', 'm s-2', conversion=US%L_T2_to_m_s2)
+  if (CS%id_hf_diffv_2d > 0) call safe_alloc_ptr(CS%hf_diffv_2d,G%isd,G%ied,G%JsdB,G%JedB)
+
```
Check 120 character line length as well

ashao commented on src/core/MOM_dynamics_split_RK2.F90:
```diff
@@ -306,6 +314,24 @@ subroutine step_MOM_dyn_split_RK2(u, v, h, tv, visc, Time_local, dt, forces, p_s
+  real, dimension(SZIB_(G),SZJ_(G)) :: hf_uBT_accel_2d ! Depth integeral of hf_uBT_accel [L T-2 ~> m s-2].
+  real, dimension(SZI_(G),SZJB_(G)) :: hf_vBT_accel_2d ! Depth integeral of hf_vBT_accel [L T-2 ~> m s-2].
+
+!  real, dimension(SZIB_(G),SZJ_(G),SZK_(G)) :: diag_hfrac_u
+!  real, dimension(SZI_(G),SZJB_(G),SZK_(G)) :: diag_hfrac_v
```
Remove commented lines

ashao commented on src/core/MOM_dynamics_split_RK2.F90:
```diff
@@ -860,6 +888,93 @@ subroutine step_MOM_dyn_split_RK2(u, v, h, tv, visc, Time_local, dt, forces, p_s
   if (CS%id_u_BT_accel > 0) call post_data(CS%id_u_BT_accel, CS%u_accel_bt, CS%diag)
   if (CS%id_v_BT_accel > 0) call post_data(CS%id_v_BT_accel, CS%v_accel_bt, CS%diag)
 
+  ! Diagnostics for terms multiplied by fractional thicknesses
+  do j=js,je ; do I=Isq,Ieq
+    hf_PFu_2d(I,j) = 0.0
+    hf_CAu_2d(I,j) = 0.0
+    hf_uBT_accel_2d(I,j) = 0.0
+  enddo ; enddo
+  do J=Jsq,Jeq ; do i=is,ie
+    hf_PFv_2d(i,J) = 0.0
+    hf_CAv_2d(i,J) = 0.0
+    hf_vBT_accel_2d(i,J) = 0.0
+  enddo ; enddo
+
```
These should be guarded behind `if (CS%id_PFu_2d>0)` flags. Also explicit loops are not required when initializing arrays.

hmkhatri replied:
Yes, I am going to add this.

ashao commented on src/core/MOM_dynamics_split_RK2.F90:
```diff
@@ -160,10 +162,16 @@ module MOM_dynamics_split_RK2
   integer :: id_umo_2d = -1, id_vmo_2d = -1
   integer :: id_PFu    = -1, id_PFv    = -1
   integer :: id_CAu    = -1, id_CAv    = -1
+  integer :: id_hfPFu  = -1, id_hfPFv  = -1
+  integer :: id_hfPFu_2d = -1, id_hfPFv_2d = -1
+  integer :: id_hfCAu  = -1, id_hfCAv  = -1
+  integer :: id_hfCAu_2d  = -1, id_hfCAv_2d  = -1
 
   ! Split scheme only.
   integer :: id_uav        = -1, id_vav        = -1
   integer :: id_u_BT_accel = -1, id_v_BT_accel = -1
+  integer :: id_hfu_BT_accel = -1, id_hfv_BT_accel = -1
+  integer :: id_hfu_BT_accel_2d = -1, id_hfv_BT_accel_2d = -1
```
These could be aligned to make them look a little neater

ashao commented on src/core/MOM_dynamics_split_RK2.F90:
```diff
@@ -860,6 +888,93 @@ subroutine step_MOM_dyn_split_RK2(u, v, h, tv, visc, Time_local, dt, forces, p_s
+    call post_data(CS%id_hfCAv, hf_CAv, CS%diag)
+  endif
+  if (CS%id_hfCAu_2d > 0) then
+    do k=1,nz ; do j=js,je ; do I=Isq,Ieq
+      hf_CAu_2d(I,j) = hf_CAu_2d(I,j) + CS%CAu(I,j,k) * CS%diag_hfrac_u(I,j,k)
+    enddo ; enddo ; enddo
+    call post_data(CS%id_hfCAu_2d, hf_CAu_2d, CS%diag)
+  endif
+  if (CS%id_hfCAv_2d > 0) then
+    do k=1,nz ; do J=Jsq,Jeq ; do i=is,ie
+      hf_CAv_2d(i,J) = hf_CAv_2d(i,J) + CS%CAv(i,J,k) * CS%diag_hfrac_v(i,J,k)
+    enddo ; enddo ; enddo
+    call post_data(CS%id_hfCAv_2d, hf_CAv_2d, CS%diag)
+  endif
+
+  if (CS%id_hfu_BT_accel > 0) then
+    do k=1,nz ; do j=js,je ; do I=Isq,Ieq
+      hf_uBT_accel(I,j,k) = CS%u_accel_bt(I,j,k) * CS%diag_hfrac_u(I,j,k)
+    enddo ; enddo ; enddo
+    call post_data(CS%id_hfu_BT_accel, hf_uBT_accel, CS%diag)
+  endif
+  if (CS%id_hfv_BT_accel > 0) then
+    do k=1,nz ; do J=Jsq,Jeq ; do i=is,ie
+      hf_vBT_accel(i,J,k) = CS%v_accel_bt(i,J,k) * CS%diag_hfrac_v(i,J,k)
+    enddo ; enddo ; enddo
+    call post_data(CS%id_hfv_BT_accel, hf_vBT_accel, CS%diag)
+  endif
+  if (CS%id_hfu_BT_accel_2d > 0) then
+    do k=1,nz ; do j=js,je ; do I=Isq,Ieq
+      hf_uBT_accel_2d(I,j) = hf_uBT_accel_2d(I,j) + CS%u_accel_bt(I,j,k) * CS%diag_hfrac_u(I,j,k)
+    enddo ; enddo ; enddo
+    call post_data(CS%id_hfu_BT_accel_2d, hf_uBT_accel_2d, CS%diag)
+  endif
+  if (CS%id_hfv_BT_accel_2d > 0) then
+    do k=1,nz ; do J=Jsq,Jeq ; do i=is,ie
+      hf_vBT_accel_2d(i,J) = hf_vBT_accel_2d(i,J) + CS%v_accel_bt(i,J,k) * CS%diag_hfrac_v(i,J,k)
+    enddo ; enddo ; enddo
+    call post_data(CS%id_hfv_BT_accel_2d, hf_vBT_accel_2d, CS%diag)
+  endif
+
```
Array variable names and id variable names should be the same: e.g. `id_hfv_BT_accel` and `hf_vBT_accel`, should be renamed to: `id_hf_vBT_accel`

ashao commented on src/core/MOM_dynamics_split_RK2.F90:
```diff
@@ -1231,6 +1353,23 @@ subroutine initialize_dyn_split_RK2(u, v, h, uh, vh, eta, Time, G, GV, US, param
   CS%id_PFv = register_diag_field('ocean_model', 'PFv', diag%axesCvL, Time, &
       'Meridional Pressure Force Acceleration', 'm s-2', conversion=US%L_T2_to_m_s2)
+  CS%id_hfPFu = register_diag_field('ocean_model', 'hf_PFu', diag%axesCuL, Time, &
+      'Thickness-weighted Zonal Pressure Force Acceleration', 'm s-2', v_extensive=.true., conversion=US%L_T2_to_m_s2)
+  CS%id_hfPFv = register_diag_field('ocean_model', 'hf_PFv', diag%axesCvL, Time, &
+      'Thickness-weighted Meridional Pressure Force Acceleration', 'm s-2', v_extensive=.true., conversion=US%L_T2_to_m_s2)
+  CS%id_hfCAu = register_diag_field('ocean_model', 'hf_CAu', diag%axesCuL, Time, &
+      'Thickness-weighted Zonal Coriolis and Advective Acceleration', 'm s-2', v_extensive=.true., conversion=US%L_T2_to_m_s2)
+  CS%id_hfCAv = register_diag_field('ocean_model', 'hf_CAv', diag%axesCvL, Time, &
+      'Thickness-weighted Meridional Coriolis and Advective Acceleration', 'm s-2', v_extensive=.true., &
+      conversion=US%L_T2_to_m_s2)
+  CS%id_hfPFu_2d = register_diag_field('ocean_model', 'hf_PFu_2d', diag%axesCu1, Time, &
+      'Barotropic Thickness-weighted Zonal Pressure Force Acceleration', 'm s-2', conversion=US%L_T2_to_m_s2)
+  CS%id_hfPFv_2d = register_diag_field('ocean_model', 'hf_PFv_2d', diag%axesCv1, Time, &
+      'Barotropic Thickness-weighted Meridional Pressure Force Acceleration', 'm s-2', conversion=US%L_T2_to_m_s2)
+  CS%id_hfCAu_2d = register_diag_field('ocean_model', 'hf_CAu_2d', diag%axesCu1, Time, &
+      'Barotropic Thickness-weighted Zonal Coriolis and Advective Acceleration', 'm s-2', conversion=US%L_T2_to_m_s2)
+  CS%id_hfCAv_2d = register_diag_field('ocean_model', 'hf_CAv_2d', diag%axesCv1, Time, &
+      'Barotropic Thickness-weighted Meridional Coriolis and Advective Acceleration', 'm s-2', conversion=US%L_T2_to_m_s2)
```
Similar to above, the variable id name and the diagnostic name should be the same.

ashao commented on src/diagnostics/MOM_diagnostics.F90:
```diff
@@ -1622,6 +1663,16 @@ subroutine MOM_diagnostics_init(MIS, ADp, CDp, Time, G, GV, US, param_file, diag
     call register_time_deriv(lbound(MIS%h), MIS%h, CS%dh_dt, CS)
   endif
 
+  CS%id_hf_du_dt = register_diag_field('ocean_model', 'hf_dudt', diag%axesCuL, Time, &
+      'Thickness-weighted Zonal Acceleration', 'm s-2', v_extensive=.true., conversion=US%L_T2_to_m_s2)
+  CS%id_hf_dv_dt = register_diag_field('ocean_model', 'hf_dvdt', diag%axesCvL, Time, &
+      'Thickness-weighted Meridional Acceleration', 'm s-2', v_extensive=.true., conversion=US%L_T2_to_m_s2)
+  CS%id_hf_du_dt_2d = register_diag_field('ocean_model', 'hf_dudt_2d', diag%axesCu1, Time, &
+      'Barotropic Thickness-weighted Zonal Acceleration', 'm s-2', conversion=US%L_T2_to_m_s2)
+  CS%id_hf_dv_dt_2d = register_diag_field('ocean_model', 'hf_dvdt_2d', diag%axesCv1, Time, &
+      'Barotropic Thickness-weighted Meridional Acceleration', 'm s-2', conversion=US%L_T2_to_m_s2)
+
+
```
Id names and diagnostic names should be the same

ashao commented on src/parameterizations/lateral/MOM_hor_visc.F90:
```diff
@@ -230,19 +236,24 @@ subroutine horizontal_viscosity(u, v, h, diffu, diffv, MEKE, VarMix, G, GV, US,
     vort_xy_dy, & ! y-derivative of vertical vorticity (d/dy(dv/dx - du/dy)) [L-1 T-1 ~> m-1 s-1]
     div_xx_dx, &  ! x-derivative of horizontal divergence (d/dx(du/dx + dv/dy)) [L-1 T-1 ~> m-1 s-1]
     ubtav         ! zonal barotropic vel. ave. over baroclinic time-step [L T-1 ~> m s-1]
+!    hf_diffu_2d   ! Depth integeral of hf_diffu [L T-2 ~> m s-2].
```
Remove commented line

ashao commented on src/parameterizations/lateral/MOM_hor_visc.F90:
```diff
@@ -230,19 +236,24 @@ subroutine horizontal_viscosity(u, v, h, diffu, diffv, MEKE, VarMix, G, GV, US,
     vort_xy_dx, & ! x-derivative of vertical vorticity (d/dx(dv/dx - du/dy)) [L-1 T-1 ~> m-1 s-1]
     div_xx_dy, &  ! y-derivative of horizontal divergence (d/dy(du/dx + dv/dy)) [L-1 T-1 ~> m-1 s-1]
     vbtav         ! meridional barotropic vel. ave. over baroclinic time-step [L T-1 ~> m s-1]
+!    hf_diffv_2d   ! Depth integeral of hf_diffv [L T-2 ~> m s-2].
```
Remove commented line

ashao commented on src/parameterizations/lateral/MOM_hor_visc.F90:
```diff
@@ -284,6 +295,9 @@ subroutine horizontal_viscosity(u, v, h, diffu, diffv, MEKE, VarMix, G, GV, US,
     grad_vel_mag_bt_q, &  ! Magnitude of the barotropic velocity gradient tensor squared at q-points [T-2 ~> s-2]
     boundary_mask_q ! A mask that zeroes out cells with at least one land edge [nondim]
 
+!  real, dimension(SZIB_(G),SZJ_(G),SZK_(G)) :: hf_diffu ! Zonal hor. visc. accel. x fract. thickness [L T-2 ~> m s-2].
+!  real, dimension(SZI_(G),SZJB_(G),SZK_(G)) :: hf_diffv ! Merdional hor. visc. accel. x fract. thickness [L T-2 ~> m s-2].
+
```
Remove commented line

ashao reviewed (changes requested):
Overall looks pretty good. Couple of general (minor comments): 
- `id_{diag_name}` and actual diagnostic names should be the same so that it's easier to search through the code
- Maximum line length is 120 characters.
- Remove commented lines

The only thing that I can imagine someone might comment on is how the fractional thicknesses get passed around. I prefer the way you've done it, but just be ready to have that conversation if someone wants it.

When you eventually submit a PR to `dev/gfdl`, you can mark me as one of the requested reviewers.

hmkhatri committed:
Shorter initlization of 2D arrays

hmkhatri committed:
Modifications in vertical friction diagnostics

Commit:
added diagnostics for partial derivative of density wrt temperature and salinity

adcroft committed:
Remove "-B" flag from pipeline "make test"

- The -B option forced make to make all targets. This was to ensure
  that everything was remade when work in a shared work space which
  is no longer the case. We always have a clean work space in these
  tests and in order to invoke make twice to handle srun errors we
  can't use the -B option anymore.

adcroft committed:
Merge branch 'more-robust-pipeline' into dev/gfdl

adcroft committed:
Merge branch 'dev/gfdl' into fix-pkg

adcroft committed:
Merge branch 'doxy-pkg' of https://github.com/MJHarrison-GFDL/MOM6 into fix-pkg

adcroft committed:
Adjusted comments in ODA hooks to fit line length

hmkhatri committed:
Diagnotics initilization as pointers to save memory allocation

marshallward committed:
Merge pull request #1139 from adcroft/fix-pkg

Moved hooks from pkg to external

marshallward committed:
Merge branch 'dev/gfdl' into use-fms-2019-01-02-in-testing

Commit:
fixes import of dens deriv function

marshallward committed:
Merge pull request #1140 from adcroft/use-fms-2019-01-02-in-testing

Use FMS tag 2019.01.02 in .testing

adcroft committed:
Merge branch 'dev/gfdl' into diag_axis_init_FMS2019.01.02

marshallward committed:
Testing: Removed environment flag checks

Due to unpredictable conflicts with flags, the realization that flag
testing would launch jobs on Slurm environment, and the fact that Slurm
launchers, OpenMPI, and MPICH all appear to pass environment variables
on default, we now just assume that "env=val $(MPIRUN)" will pass an
environment variable to the MPI job in all cases.

This may not be true on all systems, but for now we will assume this
works.

adcroft committed:
Merge pull request #1143 from nikizadehgfdl/diag_axis_init_FMS2019.01.02

This is the minimum required update to use 2020.01 FMS & FMScoupler

Hallberg-NOAA committed:
Merge branch 'dev/gfdl' into test_mpi_flag_remove

adcroft committed:
Merge pull request #1144 from marshallward/test_mpi_flag_remove

Testing: Removed environment flag checks

adcroft committed:
Merge branch 'dev/gfdl' into add_dens_deriv_diag

Hallberg-NOAA committed:
(*)Corrected halo size in EOS call if VERTEX_SHEAR=T

  Corrected halo size in density derivative calculations in smoothed_dRdT_dRdS
This fixes an i-parallelization problem that was recently introduced (as a part
of MOM6 PR#1089) when VERTEX_SHEAR is True, and closes MOM6 issue #1146.  All
answers in the existing MOM6-examples test suite are bitwise identical, but this
does change (correct) answers when VERTEX_SHEAR is true.

nikizadehgfdl committed:
Fixing openmp issues with FMS2020 cpu affinity

- FMS2020 has newer cpu affinity work. These are mostly to fix the
  issues with thread placing and  hyperthreadng under slurm on gaea.
  But it also works on Orion.
- The new affinity module simplifies the thread-placing calls in the
  component models.
- The name of some functions has changed, that's the reason for crashes
  like:
      FATAL: input domain does not have an io_domain.
- This update fixes those issues.
- openmp runs with 1 and 2 threads gives the same answers as non-openmp
- NOTE: I don't rememer why we put the thread placing calls in MOM_domains.F90
        They look as unnecessary and the whole #ifndef NOT_SET_AFFINITY block
        can probably be removed. ocean_nthreads is either set in coupler
or solo_driver.

hmkhatri committed:
Vertical friction module

adcroft committed:
Merge pull request #1147 from Hallberg-NOAA/fix_vertex_shear_halo

(*)Corrected halo size in EOS call if VERTEX_SHEAR=T

adcroft committed:
Merge branch 'dev/gfdl' into fix_openmp_affinity_issues_with_FMS2020

hmkhatri committed:
Removed commented lines and trailing spaces

kshedstrom committed:
OBC code for slope_x, slope_y.

marshallward committed:
Testing: Explicit OpenMP CPU affinity

FMS affinity operations, often used in OpenMP directives, require
explicit CPU affinities such that the number of available CPUs matches
the number of requested PEs.

To accommodate this, we explicit configure OpenMP to use cpu=0 for our
ARM tests.

Hallberg-NOAA committed:
Removed #ifdef debugging blocks

  Removed old debugging code in blocks of code surrounded by #ifdef statements
and removed unnecessary #ifdef around other blocks of debugging code.  MOM6
standards discourage the use of CPP macros except for a limited set of uses
related to memory where this is unavoidable, so this commit is bringing MOM6
closer to its stated standards.  All answers and output are identical.

kshedstrom committed:
Merge remote-tracking branch 'gfdl/dev/gfdl' into dev/esmg

Hallberg-NOAA committed:
+Aligned newlines with module documentation

  Modified doc_module so that new lines are added only when modules are
documented, and are added in all parameter_doc files in which modules are
documented.  All answers and output are identical, but there are white space
changes in MOM_parameter_doc and SIS_parameter_doc files.

hmkhatri committed:
Diagnostic description change

kshedstrom committed:
Added halo updates for OBC radiation speeds.

hmkhatri committed:
Diagnostic description change

kshedstrom committed:
Fixing the OMP issue?

kshedstrom committed:
Was hoping this would fix flux_y, but no.

hmkhatri committed:
Fractional thickness-weighted diagnostics for accelaration due to relative vorticity and gradient of kinetic energy

Hallberg-NOAA committed:
+Add like_default optional argument to log_param

  Added the new optional argument like_default to the log_param and doc_param
routines to help control where the documentation appears.  This new argument is
used for logging EPBL_USTAR_MIN, the diagnosed output value of MAXIMUM_DEPTH
when the input value is negative, and the diagnosed number of columns where
sponges occur with MOM_ALE_sponge.  An '!' was also added to the logging of
EPBL_USTAR_MIN.  All answers are bitwise identical but there are minor changes
in the contents of some MOM_parameter_doc.short files.

Commit:
altered OpenMP directive for diagnostic

Commit:
Merge branch 'add_dens_deriv_diag' of github.com:gmacgilchrist/MOM6 into add_dens_deriv_diag

marshallward committed:
Merge branch 'dev/gfdl' into add_dens_deriv_diag

marshallward committed:
Merge pull request #1142 from gmacgilchrist/add_dens_deriv_diag

Added diagnostics for partial derivatives of density

hmkhatri committed:
Modifications in vertical friction diagnostics (fractional thickness-weighted)

marshallward committed:
Merge branch 'dev/gfdl' into remove_ifdefs

marshallward committed:
Merge pull request #1152 from Hallberg-NOAA/remove_ifdefs

Removed #ifdef debugging blocks

Hallberg-NOAA committed:
Merge branch 'dev/gfdl' into module_doc_newlines

Hallberg-NOAA committed:
+Set all_default for 4 log_version calls

  Added code to determine whether all parameters in the MOM_grid, MOM_restart,
MOM_write_cputime and MOM_tracer_registry modules are being used with their
default settings, and added all_default arguments to the log_version calls for
these modules.  All answers and output are identical, but there are white space
changes in MOM_parameter_doc.short and SIS_parameter_doc.short files.

Hallberg-NOAA committed:
Merge branch 'dev/gfdl' into omp_affinity_test

Hallberg-NOAA committed:
Merge pull request #1150 from marshallward/omp_affinity_test

Testing: Explicit OpenMP CPU affinity

Hallberg-NOAA committed:
Merge branch 'dev/gfdl' into module_doc_newlines

adcroft committed:
Merge branch 'dev/gfdl' into dev/esmg

adcroft committed:
Correted OMP directives for two OBC variables

- Members of a type cannot be individually labelled as shared/private
- One variable was converted to shared since it was defiend in a non-OMP
  section and then labelled as private which meant it was uninitialized.

adcroft committed:
Merge branch 'ESMG-dev/esmg' into dev/gfdl

adcroft committed:
Merge branch 'dev/gfdl' into fix_openmp_affinity_issues_with_FMS2020

adcroft committed:
Correct scaling for second derivs in Stanley param

- Passing the scale= to calculate_density_second_derivs() was the
  double scaling the density contribution from SGS variance in the
  Stanley parameterization.

adcroft committed:
Corrected indentation in MOM_EOS

- select case statements were indented as if once inside a loop.

adcroft committed:
Fixed pressure scaling for calculate_stanley_density_array()

- The pressure scaling was wrong when trying to call
  calculate_density_second_derivs_array() from within
  calculate_stanley_density_array() because the latter should not do
  any scaling but the former always did. I had to call the lower level
  functions provided by WRIGHT, TEOS10, etc to avoid get the scaling
  tests to pass.

adcroft committed:
Merge branch 'module_doc_newlines' of https://github.com/Hallberg-NOAA/MOM6 into Hallberg-NOAA-module_doc_newlines

adcroft committed:
Merge branch 'Hallberg-NOAA-module_doc_newlines' into dev/gfdl

adcroft committed:
Fixed trailing whtiespace for #1148

adcroft committed:
Merge branch 'nikizadehgfdl-fix_openmp_affinity_issues_with_FMS2020' into dev/gfdl

adcroft committed:
Merge branch 'fix_openmp_affinity_issues_with_FMS2020' of https://github.com/nikizadehgfdl/MOM6 into nikizadehgfdl-fix_openmp_affinity_issues_with_FMS2020

adcroft committed:
Merge branch 'nikizadehgfdl-fix_openmp_affinity_issues_with_FMS2020' into dev/gfdl

adcroft committed:
Re-factored int_density_dz_generic_plm()

- In preparing to add Brankartc terms to PLM form of presure gradient
  I'm adapting the routine to look like the PPM routine which takes 3D
  arguments but only works on layer "k".

adcroft committed:
Implement Brankart terms in PLM form of PGF

- The Brankart PGF terms are now implemented in the PLM recontruction
  routines, just as they were for the PPM form.

adcroft committed:
Cleaned up tc2 MOM_input

- Added documentation lines for PGF_STANLEY_T2_DET_COEFF

hmkhatri committed:
Merge branch 'dev/gfdl' into diag_TW

hmkhatri committed:
Modifications in diag_hfrac_u and diag_hfrac_v calls

adcroft committed:
Added comments to highlight a dimensionally problematic constant

- Per feedback to #1156

adcroft committed:
Corrected scaling in _1d and _scalar EOS functions

- Per feedback on #1156, corrected scaling for calculate_stanley_density_scalar()
  and calculate_stanley_density_1d().

adcroft committed:
Removed unused member in PressureForce_CS

- Per feedback on #1156.

adcroft committed:
Added FATAL if trying to use parameterization in non-Boussinesq mode

- Per feedback on #1156

Hallberg-NOAA committed:
(*)Set maximum value of au_visc

  Set a hard-coded maximum value of CS%a_u and CS%a_v of 1e37 m s-1 so that
these can be represented in diagnostics that are written with 32-bit floating
point numbers.  These values are so large that all answers are bitwise identical
in the MOM6-examples test cases, but it is possible that answers could change.

adcroft committed:
Added comments for local variables

- Per feedback on #1156.

Hallberg-NOAA committed:
Merge branch 'adcroft-brankart' into dev/gfdl

Hallberg-NOAA committed:
+Eliminate uhbt_IC and ubt_IC from restart files

  Eliminated the unused variables CS%uhbt_IC and CS%vhbt_IC from the barotropic
control structure and from the MOM6 restart files.  Also placed a logical test
for CS%Gradual_BT_ICs around all references to CS%ubt_IC and CS%vbt_IC and
eliminated these variables from the restart files when they would not be used.
Additionally some unused arguments were removed from internal subroutines and
some spelling or index case errors were corrected.  All answers are bitwise
identical, but the MOM6 restart files have fewer variables and some unused
entries in the MOM_parameter_doc files have changed.

Hallberg-NOAA committed:
+Add the new parameter INTEGRAL_BT_CONTINUITY

  Add the new runtime parameter INTEGRAL_BT_CONTINUITY which enables the use of
the time-integrated barotropic velocity to determine the cumulative transport
since the start of the barotropic stepping.  This new option works in all of the
MOM6-examples test cases with a SPLIT=True and USE_BT_CONT_TYPE=True.  By
default all answers are bitwise identical, but there are changes to the entries
in the MOM_parameter_doc files.

Hallberg-NOAA committed:
+Add find_duhbt_dubt_int

  Added the new subroutines find_duhbt_dubt_int and find_dvhbt_dvbt_int, and use
the time-integrated forms to set the inverse face area, the transport correction
for consistency between the transport from the initial barotropic velocity and
the summed layer transports, and the maximum corrective mass source when
INTEGRAL_BT_CONT=True.  Also only log BT_CONT_CORR_BOUNDS when BOUND_BT_CORR is
set to true.  By default, all answers are bitwise identical, but there are minor
changes to the entries in the MOM_parameter_doc files.

hmkhatri committed:
Made allocation of hfrac at u/v points optional. These arrays are only allocated if any of the relevant diagnostics are called.

Hallberg-NOAA committed:
+Implemented integral_BT_cont options for OBCs

  Added code to implement new integral_BT_cont options within the barotropic
open boundary condition code.  A number of new arguments were added to
apply_velocity_OBCS.  In addition, the handling of updates to ubt_sum, uhbt_sum
and ubt_wtd with open boundary conditions were simplified.  There are minor
answer changes if INTEGRAL_BT_CONTINUITY=True, but all answers in the existing
MOM6-examples tess cases are bitwise identical.

kshedstrom committed:
+Adding a halo update for tracer reservoirs.

Hallberg-NOAA committed:
+Reuse find_uhbt with INTEGRAL_BT_CONT

  Revised the dimensions of the entries in the local_BT_Cont types when
INTEGRAL_BT_CONT=True for efficiency and to enable find_uhbt and other routines
to be used regardless of the value of INTEGRAL_BT_CONT.  Some arguments that are
no longer needed have been removed from some subroutines. All answers are
bitwise identical in the MOM6-examples test suite, but changes to the order of
arithmetic leads to changes from the previous version when
INTEGRAL_BT_CONT=True, and the rescaling is controlled via new optional
arguments to internal routines.

Hallberg-NOAA committed:
Merge branch 'dev/gfdl' into integral_bt_solver

Hallberg-NOAA committed:
(*?)Revised ice_shelf_driver.F90 so it compiles

  Revised ice_shelf_driver.F90 so that it compiles successfully and is similar
to MOM_driver.F90 where possible.  In addition, deleted the files
MOM_surface_forcing.F90 and user_surface_forcing.F90 in the ice_solo_driver
directory, as these are unused and out-of-date versions of the equivalent files
in the solo_driver directory.  Some of the changes are not as streamlines as
they could be if the solo_ice_shelf code used its own version of diag_mediator,
or if the interface to diag_mediator were revised so that some of its
ocean-specific arguments are optional.  There is no test-case exercising this
code, but at least it now compiles, whereas it did not compile before.

hmkhatri committed:
Merge branch 'dev/gfdl' into diag_TW

adcroft committed:
Move FMS tag to 2019.01.03

- FMS tag 2019.01.03 fixes problems for EMC allowing the newer FV3 to
  work with MOM6 which is not yet compatible with FMS2020.

Hallberg-NOAA committed:
Avoid using uninitialized arrays

  Modified recent additions to avoid using uninitialized arrays.  This was not
detected in tests of the previous PR because the arrays that were created from
uninitialized arrays were themselves never used, so there are no answer changes
but there were errors with intolerant compiler settings.  Also corrected the
dOxygen syntax in a comment and corrected some openMP directives that had been
triggering warnings or errors.  All answers are bitwise identical.

marshallward committed:
Merge pull request #1164 from adcroft/fms-2019.01.03

Move FMS tag to 2019.01.03

hmkhatri committed:
Fractional-thickness weighted 3D diagnostics are now commented out as we do not the proper grid remapping option.

Hallberg-NOAA committed:
Merge pull request #1159 from ESMG/dev/esmg

+Adding a halo update for tracer reservoirs.

Hallberg-NOAA committed:
Added missing openMP directives

  Added some missing openMP directives that had been triggering warnings or
errors, and could have created problems with more extensive testing.  Also set
some unset array bounds in the ice_shelf_dynamics code.  All answers are bitwise
identical in all working test cases.

Hallberg-NOAA committed:
Merge branch 'dev/gfdl' into integral_bt_solver

adcroft committed:
Merge pull request #1161 from Hallberg-NOAA/integral_bt_solver

MOM6: +(*)Add option to use time-integrated barotropic continuity

adcroft committed:
Merge pull request #1162 from Hallberg-NOAA/ice_shelf_driver

MOM6: (*?)Revised ice_shelf_driver.F90 so it compiles

Commit:
Fixing loop bound error in MOM_PressureForce_Montgomery.F90

The "for" loop at L510 should be to nz instead of nz+1.  Due to array allocation sizes, this loop was extending beyond the shope of the argument arrays.

marshallward committed:
Testing: tc4 installs python-netCDF4 via venv (#1166)

Squash merge of following commits:

* Testing: tc4 installs python-netCDF4 via venv

Currently, users are expected to have numpy and netcdf4 python modules
in order to generate the necessary netCDF input files.  This fails in
environments where these modules are unavailable.

This patch now installs the modules into a virtual environment which are
accessible when generating the tc4 inputs.

This solution is local to tc4 but could be extended to other tests as
needed.

* Testing: remove Python numpy and netCDF4 modules

The numpy and netCDF4 packages are no longer needed since tc4 now
installs these locally.

* Test: Install virtualenv for Python2

tc4 was using the venv module which appears to be python3-specific, and
Travis Ubuntu defaults to python2.  Also virtualenv was not installed in
either case.

This patch adds python-virtualenv to the install packages and uses the
virtualenv module.

* Test: Add python-dev for ARM64 numpy

Travis ARM nodes need to build numpy natively when installed by pip, and
thus require Python headers.  These are provided by the python-dev
package.

* Test: Revert to Py3 for tc4 generation scripts

Reverting the python 2 support (default for Travis) to use Python 3
syntax.

The main reason is that Python 3 includes venv (equivalent to
virtualenv) as its standard library, and is therefore guaranteed to
exist if Python 3 exists.

Python 3's virtualenv must be independently installed, which cannot be
confirmed.

This will cause problems for people without Python 3, but this is
probably the best solution, or at least the starting point for a more
general solution.

* Test: Adding python3-venv to Travis

Ubuntu apparently requires an explicit install of python3-venv despite
it being part of the standard library.  Go figure...

* Test: Arm64 tc4 Configuration support

Arm64 Ubuntu environments require explicit installations which are
otherwise provided on x86 Ubuntu:

* Python 3 Pip must be installed (python3-pip)
* Wheel installation must be explicitly installed
* Cython is required for numpy
* Numpy must be explicitly built before installing python-netCDF4

* Test: tc4 common Python local-env

Build times for setting up the virtual environments can be very
expensive on the Arm64 Ubuntu nodes, so we now create a shared directory
for launching the environments.

* Test: Use '.' in place of 'source' for Make

* Test: Test for required Python modules in tc4

We have reworked the Makefile to conditionally test for required Python
modules in tc4.  If unavailable, we install these in a virtual
environment.

This does not address many scenarios, such as if Python 3 is missing,
venv is missing (as in Ubuntu), or handle the situation if they do not
exist.  It assumes that either the modules exist, or that they can be
installed by venv.

This should be seen as an iterative step to get things working on Travis
x86 and Arm64, as well as GFDL's Gaea and most user Linux platforms.

* Test: Explicit python execs for tc4 input build

This resolves some issues with python2/3 resolution and limited support
of various platforms for module support.

Specifically, older platform with basic Python 3 support may not also
have numpy support.  In this case, we can defer back to Python 2 (or
whatever the system Python may be).

* Test: Setup Python venv at build time

This patch moves the Python virtual environment configuration to the
main Makefile, which is setup at build time, rather than in the model
configuration Makefile, which will typically not have internet access if
run on a compute node.

As before, the venv will only be setup when the numpy and netCDF4
modules are unavailable.  A minor bug in the logic of the check has
also been fixed.

marshallward committed:
Merge pull request #1169 from breichl/BugFix_MontgomeryPressureForce

Fixing loop bound error in MOM_PressureForce_Montgomery.F90

hmkhatri committed:
Initilization of 2D diagnostic arrays changed from pointer stype to allocatable array style.

hmkhatri committed:
Trailing spaces removed and line lengths reduced to be under 120 characters

hmkhatri committed:
Comment modified

adcroft committed:
Merge branch 'dev/gfdl' into diag_TW
